cranelift: arm32 codegen

jmkrauz opened the pull request:
This commit adds arm32 Thumb2 code generation for some IR insts.
I have used machinst backend and followed Aarch64 implementation as an example.

Floating-point instructions are not supported, because regalloc
does not seem to allow to represent overlapping register classes,
which are needed by VFP/Neon.

Another lacking feature is support for I64 and I128 types.

An abi implementation is not complete, however it makes it possible
to run simple clif tests.

It may be a first step towards  #1173.

This PR contains instruction emission test and clif tests for IR insts.

I don't know well who could review this.

<!--
Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

bjorn3 commented on cranelift/codegen/src/regalloc/solver.rs:
```diff
@@ -1149,24 +1149,14 @@ mod tests {
-    }
+    // Arm32 `TargetIsa` is now `TargetIsaAdapter`, which does not hold any info
+    // about registers, so we directly access `INFO` from registers-arm32.rs.
+    include!(concat!(env!("OUT_DIR"), "/registers-arm32.rs"));
```
Maybe use x86 or riscv instead?

jmkrauz replied:
I believe some of the tests here or in regalloc/pressure.rs use the fact that S, D and Q registers overlap. I am not sure if we can achieve something similar with x86 or risc-v registers.

cfallin commented:
Wow -- this is an excellent surprise to start my day with. First of all, thanks, @jmkrauz, for putting in all of this work!

I'll review this in the next few days; I expect it will take a bit of time. One high-level comment regarding the register-overlap issue: we've thought about (and need to solve) it for other architectures too, e.g. x86-32, so I expect we'll have a better answer eventually. For now, a stopgap solution (for correctness) might be to simply allocate pairs as the fundamental unit (i.e. `D` regs), and just burn an extra 32-bit register for each `f32`, and punt on 128-bit vector support for now. Not ideal but it will get us to basic completeness at least. @julian-seward1 might have more thoughts on this?

bnjbvr commented:
Agreed that we could review and merge this behind a flag, or just letting know that support for arm32 is experimental; we've done this for other backends, and we can work incrementally on supporting more opcodes etc.

cfallin commented:
Hi @jmkrauz, it looks like you've refactored to use the common `abi_impl` machinery (thanks!). I discussed this PR briefly today with a few others and we agree that it might be best to merge now, given that I haven't resolved the multiple-register-value question yet (which this would need for `i64` values). (Sorry about that -- a lot on my plate lately!)  If that's alright with you, of course; does this seem OK or do you have plans to refine the PR more first?

One thing I think we should do, however, is to put the new backend behind a feature flag until it implements Wasm MVP; otherwise someone might try to build wasmtime on arm32 and hit confusing panics. Perhaps just `#[cfg(feature = "arm32")]` on the main `mod arm32` statement in `cranelift/codegen/src/isa/mod.rs`?

jmkrauz commented:
Merging now is absolutely fine with me.
I have added additional commit in which I put ARM backend behind `experimental_arm32` feature flag. In order to avoid failing arm32 filetests with `experimental_arm32` disabled, I have also added small check in `cranelift/filetests/src/runone.rs`.

cfallin commented:
Excellent, thanks! Looking forward to seeing this become a full backend!
